### PR TITLE
Bug Fix: GitHub Usernames are Case Insensitive

### PR DESCRIPTION
**Pull request by megan07**
Addressing #196 where GitHub usernames were being treated as case-sensitive in the diff.
Before fix:
```
Step 1 error: After applying this step, the plan was not empty:

        DIFF:

        DESTROY/CREATE: github_team_membership.test_team_membership
          etag:     "W/\"52d84b121c000efbcf03e7e877c4fbf7\"" => "<computed>"
          id:       "3289152:Mbang-github2" => "<computed>"
          role:     "member" => "member"
          team_id:  "3289152" => "3289152"
          username: "mbang-github2" => "Mbang-github2" (forces new resource)
```
```
Step 1 error: After applying this step, the plan was not empty:

        DIFF:

        DESTROY/CREATE: github_membership.test_org_membership
          etag:     "W/\"ef0b40d063d184967e965b7d8af83466\"" => "<computed>"
          id:       "mbang-github-acc-test:mbang-github2" => "<computed>"
          role:     "member" => "member"
          username: "mbang-github2" => "Mbang-github2" (forces new resource)
```
```
    Error: Provider produced inconsistent result after apply

        When applying changes to
        github_repository_collaborator.test_repo_collaborator, provider "github"
        produced an unexpected new value for was present, but now absent.

        This is a bug in the provider, which should be reported in the provider's own
        issue tracker.

FAIL
```

After fix:
```
=== RUN   TestAccGithubTeamMembership_basic
--- PASS: TestAccGithubTeamMembership_basic (13.25s)
PASS
```
```
=== RUN   TestAccGithubMembership_basic
--- PASS: TestAccGithubMembership_basic (4.08s)
PASS
```
```
=== RUN   TestAccGithubRepositoryCollaborator_basic
--- PASS: TestAccGithubRepositoryCollaborator_basic (10.31s)
PASS
```

I also tested that if the user DOES update the casing in the GitHub UI, a terraform refresh will pick that up.

### CODE REVIEW

File: github/resource_github_membership_test.go [hunk 1]
```diff
@@ -29,6 +30,27 @@ func TestAccGithubMembership_basic(t *testing.T) {
 	})
 }
 
+func TestAccGithubMembership_caseInsensitive(t *testing.T) {
+	if len(testCollaborator) == 0 {
```
Comment (nywilken) [inline github/resource_github_membership_test.go]: Nit when checking for empty strings its best to explicitly check for an empty string so that future readers know the variable is a string and not some type that implements the length method. 

```suggestion
	if testCollaborator == "" {
```

File: github/resource_github_membership_test.go [hunk 2]
```diff
@@ -140,3 +162,18 @@ var testAccGithubMembershipConfig string = fmt.Sprintf(`
     role = "member"
   }
 `, testCollaborator)
+
+func testAccGithubMembershipConfig_caseInsensitive() string {
+	otherCase := []rune(testCollaborator)
```
Comment (nywilken) [inline github/resource_github_membership_test.go]: If I get the intent right, you are modifying the collaborator username, by toggling the case of the first rune, to test case insensitivity functionality. Is that correct? If so what about changing the function signature to take in an input, and just pass the collaborator name with the expected case. 

```go
func testAccGithubMembershipConfig_caseInsensitive(memberName string) {
return fmt.Sprintf(`
  resource "github_membership" "test_org_membership" {
    username = "%s"
    role = "member"
  }
`, memberName)
}
```

This way in the acceptance test you can run two TestSteps:

```go
Steps: []resource.TestStep{
			{
				Config: testAccGithubMembershipConfig_caseInsensitive(strings.ToUpper(testCollaborator)),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckGithubMembershipExists("github_membership.test_org_membership", &membership),
				),
			},
			{
				Config: testAccGithubMembershipConfig_caseInsensitive(strings.ToLower(testCollaborator)),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckGithubMembershipExists("github_membership.test_org_membership", &membership),
				),
			},
		},
```

File: github/resource_github_repository_collaborator.go [hunk 3]
```diff
@@ -112,14 +114,14 @@ func resourceGithubRepositoryCollaboratorRead(d *schema.ResourceData, meta inter
...
-			if *c.Login == username {
+			if strings.ToLower(*c.Login) == strings.ToLower(username) {
 				permissionName, err := getRepoPermission(c.Permissions)
 				if err != nil {
 					return err
 				}
 
 				d.Set("repository", repoName)
-				d.Set("username", username)
+				d.Set("username", *c.Login)
```
Comment (nywilken) [inline github/resource_github_repository_collaborator.go]: You don't need to deference the pointer here, as the `d.Set` will check that it is a pointer and get the value if the pointer is not nil.

**Review by nywilken — COMMENTED** [review]
@megan07 this looks good so far. I took a quick pass and left a couple of comments. I'll circle back later today.

**Review by nywilken — COMMENTED** [review]
@megan07 I left a few comments for you, but this is otherwise good. 

```
--- PASS: TestAccGithubMembership_basic (3.30s)
--- PASS: TestAccGithubMembership_caseInsensitive (2.84s)
--- PASS: TestAccGithubMembership_importBasic (2.79s)
--- PASS: TestAccGithubRepositoryCollaborator_basic (9.65s)
--- PASS: TestAccGithubRepositoryCollaborator_importBasic (9.17s)
```

File: github/resource_github_membership_test.go [hunk 4]
```diff
@@ -2,29 +2,52 @@ package github
...
+	"unicode"
 
 	"github.com/google/go-github/v25/github"
 	"github.com/hashicorp/terraform/helper/resource"
 	"github.com/hashicorp/terraform/terraform"
 )
 
 func TestAccGithubMembership_basic(t *testing.T) {
+	if testCollaborator == "" {
+		t.Skip("Skipping because length of `GITHUB_TEST_COLLABORATOR` is not set")
```
Comment (paultyng) [inline github/resource_github_membership_test.go]: think this is a copy/paste typo for the message

File: github/resource_github_membership_test.go [hunk 5]
```diff
@@ -134,9 +157,21 @@ func testAccCheckGithubMembershipRoleState(n string, membership *github.Membersh
...
     username = "%s"
     role = "member"
   }
-`, testCollaborator)
+`, username)
+}
+
+func testAccGithubMembershipTheSame(orig, other *github.Membership) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		if *orig.URL != *other.URL {
```
Comment (paultyng) [inline github/resource_github_membership_test.go]: should this also test to make sure at least one isn't null/empty? or should two empties be equal?

File: github/resource_github_membership_test.go [hunk 6]
```diff
@@ -2,29 +2,52 @@ package github
...
 func TestAccGithubMembership_basic(t *testing.T) {
+	if testCollaborator == "" {
+		t.Skip("Skipping because length of `GITHUB_TEST_COLLABORATOR` is not set")
+	}
+
 	var membership github.Membership
+	var otherMembership github.Membership
+
+	oc := []rune(testCollaborator)
+	if unicode.IsUpper(oc[0]) {
```
Comment (paultyng) [inline github/resource_github_membership_test.go]: not sure if its possible, but could `0` ever be a number or symbol in a username?

**Review by paultyng — APPROVED** [review]
Some minor comments, but otherwise LGTM

File: github/resource_github_membership_test.go (hunk 5 above)
Comment (megan07) [reply github/resource_github_membership_test.go]: I think, for the intent of the test, two empties would be equal.  I can't think of a time when this would be empty here, but if they were to both be empty, I think it means the case change didn't have an effect.

**Review by paultyng — APPROVED** [review]

**Comment by kjolley-zainar** [comment]
I'm still seeing this:

```
% tf --version
Terraform v1.6.6
on darwin_amd64

% tf plan
...
# module.github.github_repository_collaborators.myenv will be updated in-place
  ~ resource "github_repository_collaborators" "myenv" {
        id             = "myenv"
      ~ invitation_ids = {} -> (known after apply)
        # (1 unchanged attribute hidden)

      - user {
          - permission = "pull" -> null
          - username   = "foo" -> null
        }
      - user {
          - permission = "pull" -> null
          - username   = "bar" -> null
        }
      + user {
          + permission = "pull"
          + username   = "Foo"
        }
      + user {
          + permission = "pull"
          + username   = "Bar"
        }

        # (4 unchanged blocks hidden)
    }
```

**Comment by esacteksab** [comment]
I'm still seeing this as well. Not sure if there is a regression. An example (a bit redacted).

```terraform
Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # github_team_members.drupal_fe will be updated in-place
  ~ resource "github_team_members" "drupal_fe" {
        # (2 unchanged attributes hidden)

      - members {
          - role     = "maintainer" -> null
          - username = "username" -> null
        }
      + members {
          + role     = "maintainer"
          + username = "UserName"
        }

        # (3 unchanged blocks hidden)
    }

  # github_team_members.org_admin will be updated in-place
  ~ resource "github_team_members" "org_admin" {
        # (2 unchanged attributes hidden)

      - members {
          - role     = "maintainer" -> null
          - username = "username" -> null
        }
      + members {
          + role     = "maintainer"
          + username = "UserName"
        }

        # (4 unchanged blocks hidden)
    }

  # github_team_members.org_devops will be updated in-place
  ~ resource "github_team_members" "org_devops" {
        # (2 unchanged attributes hidden)

      - members {
          - role     = "maintainer" -> null
          - username = "username" -> null
        }
      + members {
          + role     = "maintainer"
          + username = "UserName"
        }

        # (1 unchanged block hidden)
    }

  # github_team_members.org_eng will be updated in-place
  ~ resource "github_team_members" "org_eng" {
        # (2 unchanged attributes hidden)

      - members {
          - role     = "maintainer" -> null
          - username = "username" -> null
        }
      + members {
          + role     = "maintainer"
          + username = "UserName"
        }

        # (6 unchanged blocks hidden)
    }

Plan: 0 to add, 4 to change, 0 to destroy.
```
